### PR TITLE
Create PSQL_SLAVE and REP_* for streaming replication and snapshots.

### PR DESCRIPTION
**Pull request by Downchuck**
This adds replication flags for streaming replication and snapshots. It adds a flag for SSL though the end-user would need to add their own certificates. This stops short of providing archive mode flags.

### CODE REVIEW

**Comment by sameersbn** [comment]
@Downchuck Please allow me some time to try this out before merging the PR. Thanks.

**Comment by Downchuck** [comment]
At present to test on the local machine you have to run docker inspect to get the IP address of the master; I'm not sure if there are better params to use, such as --link, when testing.

In practice, a replication slave will be on a different machine and will thus need the IP through other means. It's possible that a snapshot however would reside on the same machine.

**Comment by Downchuck** [comment]
Checking on status, let me know if I should make any changes? I'll be looking into SSL soon.

**Comment by sameersbn** [comment]
Hi,

I am extremely sorry for not testing this out yet.
Please allow me until tomorrow.

Thanks

**Comment by Downchuck** [comment]
I'm not happy with the README example of replication, as the host will be based on the IP of the instances, it will not be localhost. There are the usual opportunities for linking containers and such, and such an example may be better.

As far as I can tell, the only reason to run replication on the same machine would be for snapshots, not for live replication.

File: start [hunk 1]
```diff
@@ -45,18 +61,72 @@ cat >> ${PG_CONFDIR}/pg_hba.conf <<EOF
...
-  PG_OLD_VERSION=$(find ${PG_HOME}/[0-9].[0-9]/main -maxdepth 1 -name PG_VERSION 2>/dev/null | sort -r | head -n1 | cut -d'/' -f5)
+  if [ ! "${PSQL_SLAVE}" == "false" ]; then
+   echo "Replicating database..."
+   echo "${REP_PASS}" | sudo -u postgres -H "${PG_BINDIR}/pg_basebackup" -D "${PG_DATADIR}" \
+   -h "${REP_HOST}" -p "${REP_PORT}" -U "${REP_USER}" -W -v -P
+   echo "Setting up hot standby configuration..."
+   cat >> ${PG_CONFDIR}/postgresql.conf <<EOF
+hot_standby = on
+EOF
+   cat >> ${PG_DATADIR}/recovery.conf <<EOF
```
Comment (sameersbn) [inline start]: you might want to change the ownership of the `recovery.conf` file, its currently owned by root

File: start [hunk 2]
```diff
@@ -45,18 +61,72 @@ cat >> ${PG_CONFDIR}/pg_hba.conf <<EOF
...
+
 cd ${PG_HOME}
 
 # initialize PostgreSQL data directory
 if [ ! -d ${PG_DATADIR} ]; then
-  # check if we need to perform data migration
-  PG_OLD_VERSION=$(find ${PG_HOME}/[0-9].[0-9]/main -maxdepth 1 -name PG_VERSION 2>/dev/null | sort -r | head -n1 | cut -d'/' -f5)
+  if [ ! "${PSQL_SLAVE}" == "false" ]; then
+   echo "Replicating database..."
+   echo "${REP_PASS}" | sudo -u postgres -H "${PG_BINDIR}/pg_basebackup" -D "${PG_DATADIR}" \
```
Comment (sameersbn) [inline start]: @Downchuck  The user is prompted to enter the `REP_PASS` here. The echo method doe not work. Replace command with the following to resolve the issue.

```bash
sudo -u postgres -H \
    PGPASSWORD=$REP_PASS "${PG_BINDIR}/pg_basebackup" -D "${PG_DATADIR}" \
      -h "${REP_HOST}" -p "${REP_PORT}" -U "${REP_USER}" -w -v -P
```

File: start [hunk 3]
```diff
@@ -45,18 +61,72 @@ cat >> ${PG_CONFDIR}/pg_hba.conf <<EOF
...
+hostssl replication     $REP_USER       0.0.0.0/0               md5
+EOF
+  fi
+fi
+
+
+# allow passwordless connections from current user
+if [ -n "${DB_USER}" ]; then
+  if [ ! -z "${DB_PASS}" ]; then
+    cat >> ~/.pgpass <<EOF
```
Comment (sameersbn) [inline start]: @Downchuck If this change is not related to adding replication support, then I request you to submit an separate PR with this change. It will allow me to merge such changes immediately and/or initiate a separate discussion regarding it.Try and limit your PR on only to the replication feature. I hope you understand.

File: start [hunk 4]
```diff
@@ -82,6 +152,25 @@ if [ -n "${PG_OLD_VERSION}" ]; then
     -O "-c config_file=${PG_CONFDIR}/postgresql.conf" >/dev/null
 fi
 
+
+# Hot standby servers can ignore the following code.
+if [ "${PSQL_SLAVE}" == "false" ]; then
+
+if [ -n "${REP_USER}" ]; then
```
Comment (sameersbn) [inline start]: @Downchuck Need to fix the indentation levels in this block.

File: start [hunk 5]
```diff
@@ -45,18 +61,72 @@ cat >> ${PG_CONFDIR}/pg_hba.conf <<EOF
...
+EOF
+  else
+    # check if we need to perform data migration
+    PG_OLD_VERSION=$(find ${PG_HOME}/[0-9].[0-9]/main -maxdepth 1 -name PG_VERSION 2>/dev/null | sort -r | head -n1 | cut -d'/' -f5)
 
-  echo "Initializing database..."
-  sudo -u postgres -H "${PG_BINDIR}/initdb" --pgdata="${PG_DATADIR}" \
+    echo "Initializing database..."
+    sudo -u postgres -H "${PG_BINDIR}/initdb" --pgdata="${PG_DATADIR}" \
     --username=postgres --encoding=unicode --auth=trust >/dev/null
```
Comment (sameersbn) [inline start]: @Downchuck in this case and others, please indent the continuation of the command when it is continued on the second line.

File: README.md [hunk 6]
```diff
@@ -143,6 +144,45 @@ This has the effect of adding the following to the `pg_hba.conf` file:
...
+Create a slave instance with streaming replication:
+
+```bash
+docker run --name postgresql-slave -p 5433:5432 -d \
+  -e 'PSQL_TRUST_LOCALNET=true' \
+  -e 'PSQL_SLAVE=true' \
+  -e 'REP_HOST=localhost' -e 'REP_PORT=5432' \
+  -e 'REP_USER=replicator' -e 'REP_PASS=replicatorpass' \
+  sameersbn/postgresql:9.4
+```
```
Comment (sameersbn) [inline README.md]: @Downchuck Maybe we can use docker links in the the example command to fill out the `REP_HOST` variable. 

```bash
docker run --name postgresql-slave -d \
  --link postgresql-master:postrgresql \
  -e 'PSQL_TRUST_LOCALNET=true' \
  -e 'PSQL_SLAVE=true' \
  -e 'REP_HOST=postgresql' -e 'REP_USER=replicator' -e 'REP_PASS=replicatorpass' \
  sameersbn/postgresql:9.4
```

Will indirectly configure the `REP_HOST` on the slave to the `postgresql-master` containers ip address. This way in the example command to start the master we don't have to specify the `-p 5432:5432` option. That can be left up to the user to customize (user's seem to take such commands too literally).

File: README.md [hunk 7]
```diff
@@ -143,6 +144,45 @@ This has the effect of adding the following to the `pg_hba.conf` file:
...
+
+Create a snapshot downloaded from a master instance:
+```bash
+docker run --name postgresql-slave -p 5433:5432 -d \
+  -e 'PSQL_TRUST_LOCALNET=true' \
+  -e 'PSQL_SLAVE=snapshot' \
```
Comment (sameersbn) [inline README.md]: @Downchuck What is the difference between `PSQL_SLAVE=true` and `PSQL_SLAVE=snapshot`? upon execution they seem to do the exact same thing. In the code I was not able to figure out the difference.

**Comment by sameersbn** [comment]
@Downchuck Let me start off by saying this is an excellent feature. So thank you.

In addition to the code review, here are my overall comments.
 - Change the `PSQL_SLAVE` option to `PSQL_MODE` where it can be set to `master`, `slave` or `snapshot`. It would be set to `master` by default. Feel free to disagree :smile: 
 - I think we need to change the `REP_` prefix the the replication options to something more intuitive. I am not sure what to use. If nothing works then we can just use `REPLICATION_`.

With regard to you comment about using docker links to automatically fetch the `REP_` parameters, It is good to have this support. But this can be handled in a separate PR. Lets limit this PR to only adding replication support.

Finally, let me apologize for the delay in reviewing the PR. I had too much on my plate at the time. Please expect quicker responses since I have familiarized myself with the changes.

Thank you.

**Comment by sameersbn** [comment]
@Downchuck Please rebase the PR when updating.

**Comment by Downchuck** [comment]
Great code review! Thanks so much. I'll get you an updated PR next week.

**Comment by sameersbn** [comment]
@Downchuck FWIW, here are the changes I made while testing locally https://dl.dropboxusercontent.com/u/45559595/changeset.patch. Note this patch will only apply cleanly after you rebase your PR.

For testing I used the following commands.

```bash
# master
docker run --name='postgresql' -it --rm \
  --volume=/tmp/postgresql-master:/var/lib/postgresql \
  -e 'PSQL_TRUST_LOCALNET=true' \
  -e 'REP_USER=replicator'  -e 'REP_PASS=replicatorpass' \
  sameersbn/postgresql:latest

# slave
docker run --link postgresql:psql --name='postgresql-slave' -it --rm \
  --volume=/tmp/postgresql-slave:/var/lib/postgresql \
  -e 'PSQL_TRUST_LOCALNET=true' \
  -e 'PSQL_SLAVE=true' \
  -e 'REP_HOST=psql' -e 'REP_USER=replicator' -e 'REP_PASS=replicatorpass' \
  sameersbn/postgresql:latest
```
One thing I forgot to note, in the README we should indicate that this feature is only available in image versions ` > 9.4`

File: start (hunk 3 above)
Comment (Downchuck) [inline start]: Issue #23

File: README.md (hunk 7 above)
Comment (Downchuck) [inline README.md]: A snapshot is just a backup, it is not maintained through streaming replication but is only a snapshot created when it's booted up.

**Comment by Downchuck** [comment]
Rebase didn't go quite how I wanted, let me know if you need me to run another one.

**Comment by Downchuck** [comment]
Well, apart from that frustrating rebase: I took your input and applied all of it. Using --link during development made things very easy. PSQL_MODE is now "standalone" by default, as that mode does not have replication, then there's "master", "slave" and "snapshot", all of which appear to work. It's also got ARCHIVEMODE to stash WAL archives for extended slave/snapshot function.

**Comment by Downchuck** [comment]
@sameersbn - Checking in. I'd like to put in some work on testing SSL-enabled replication and I want it to be based on this patch having been applied.

**Comment by sameersbn** [comment]
@Downchuck Hi. I am testing the changes.

File: README.md [hunk 8]
```diff
@@ -143,6 +144,68 @@ This has the effect of adding the following to the `pg_hba.conf` file:
...
+# Creating a Snapshot or Slave Database
+
+You may use the `PSQL_MODE` variable along with `REPLICATION_HOST`, `REPLICATION_PORT`, `REPLICATION_USER` and `REPLICATION_PASS` to create a snapshot of an existing database and enable stream replication.
+
+Your master database must support replication or super-user access for the credentials you specify. The `PSQL_MODE` variable should be set to "master" for replication on your master node and "slave" or "snapshot" for streaming replication or a point-in-time snapshot of a running database.
+
+Create a new master instance:
+
+```bash
+docker run --name postgresql-master 5432:5432 -d \
```
Comment (sameersbn) [inline README.md]: @Downchuck You missed the `-p` argument here.

File: README.md [hunk 9]
```diff
@@ -143,6 +144,68 @@ This has the effect of adding the following to the `pg_hba.conf` file:
...
+  -e 'DB_NAME=dbname' \
+  -e 'DB_USER=dbuser' -e 'DB_PASS=dbpass' \
+  -e 'REPLICATION_USER=replicator'  -e 'REPLICATION_PASS=replicatorpass' \
+  sameersbn/postgresql:9.4
+```
+
+Create a slave instance with streaming replication:
+
+```bash
+docker run --name postgresql-slave -p 5433:5432 -d \
```
Comment (sameersbn) [inline README.md]: This command does not work because of `-e 'REPLICATION_HOST=localhost'`. You can instead do `--link postgresql-master:master -e 'REPLICATION_HOST=master`. Here `master` will be point to the `postgresql-master` container in the `/etc/hosts` file.

After making the above changes in the test command, the replication works, here is the output

```bash
Enabling trust samenet in pg_hba.conf...
Replicating database...
26665/26665 kB (100%), 1/1 tablespace                                         
NOTICE:  WAL archiving is not enabled; you must ensure that all required WAL segments are copied through other means to complete the backup
pg_basebackup: base backup completed
Setting up hot standby configuration...
Starting PostgreSQL server...
2015-06-25 05:31:15 UTC [23-1] LOG:  database system was interrupted; last known up at 2015-06-25 05:31:15 UTC
2015-06-25 05:31:15 UTC [23-2] LOG:  entering standby mode
```

But after that the following message is printed repeatedly.

```bash
FATAL:  could not connect to the primary server: invalid sslmode value: "dbname=replication"
```

Is this the expected behaviour?

File: README.md [hunk 10]
```diff
@@ -143,6 +144,68 @@ This has the effect of adding the following to the `pg_hba.conf` file:
...
+  sameersbn/postgresql:latest
+
+# slave
+docker run --link postgresql-master:psql --name='postgresql-slave' -it --rm \
+  --volume=/tmp/postgresql-slave:/var/lib/postgresql \
+  -e 'PSQL_TRUST_LOCALNET=true' \
+  -e 'PSQL_MODE=slave' \
+  -e 'REPLICATION_HOST=psql' -e 'REPLICATION_USER=replicator' -e 'REPLICATION_PASS=replicatorpass' \
+  sameersbn/postgresql:latest
+```
```
Comment (sameersbn) [inline README.md]: I tested these commands, logs indicate that the backup is being generated

```bash
Enabling trust samenet in pg_hba.conf...
Replicating database...
20129/20129 kB (100%), 1/1 tablespace
NOTICE:  WAL archiving is not enabled; you must ensure that all required WAL segments are copied through other means to complete the backup
pg_basebackup: base backup completed
Setting up hot standby configuration...
Starting PostgreSQL server...
2015-06-25 05:58:28 UTC [22-1] LOG:  database system was interrupted; last known up at 2015-06-25 05:58:28 UTC
2015-06-25 05:58:28 UTC [22-2] LOG:  entering standby mode
```

But then I see the following line printed over and over

```bash
FATAL:  could not connect to the primary server: invalid sslmode value: "dbname=replication"
```

Is this the expected behaviour?

Comment (Downchuck) [inline README.md]: That sslmode error is not expected; it's a bad variable assignment somewhere.

**Comment by Downchuck** [comment]
I think I had a bad merge or a poor oversight: PSQL_SSLMODE and the rest of the variables should be set to default values at the head of the shell script. PSQL_SSLMODE=${PSQL_SSLMODE:"disable"}; primarily.

File: start [hunk 11]
```diff
@@ -60,18 +62,78 @@ cat >> ${PG_CONFDIR}/pg_hba.conf <<EOF
...
+if [ "${PSQL_MODE}" == "master" ]; then
+  if [ -n "${REPLICATION_USER}" ]; then
+    echo "Supporting hot standby..."
+    cat >> ${PG_CONFDIR}/postgresql.conf <<EOF
+wal_level = hot_standby
+max_wal_senders = 3
+checkpoint_segments = 8
+wal_keep_segments = 8
+EOF
+    if [ "${PSQL_ARCHIVEMODE}" == "on" ]; then
```
Comment (sameersbn) [inline start]: @Downchuck I believe the `PSQL_ARCHIVEMODE` change is not related to the replication feature right? I suggest this be submitted as a separate PR for reasons mentioned earlier.

However I tried adding `-e 'PSQL_ARCHIVEMODE=on'` in the launch command of the master and got these errors.

```bash
Enabling trust samenet in pg_hba.conf...
Supporting hot standby...
Creating user "replicator"...
LOG:  unrecognized configuration parameter "archive_cleanup_command" in file "/etc/postgresql/9.4/main/postgresql.conf" line 622
FATAL:  configuration file "/etc/postgresql/9.4/main/postgresql.conf" contains errors
```

**Comment by sameersbn** [comment]
@Downchuck yes.. I remember it worked fine the last time. Let me know when you update the PR so I can retest.

P.S. I think we should set `REPLICATOR_PORT` to `5432` by default, i.e. `REPLICATOR_PORT=${REPLICATOR_PORT:-5432}`

**Comment by Downchuck** [comment]
@sameersbn PSQL_ARCHIVEMODE is used for replication, though it's an optional feature.

**Comment by Downchuck** [comment]
@sameersbn If you do see any changes that you'd like made in the pull request, you are certainly free to add them to the commit/merge.

**Comment by sameersbn** [comment]
@Downchuck did you update the PR to fix the issues i reported last time?

**Comment by Downchuck** [comment]
@sameersbn Yes - my merge went awrong: I missed the chunk of code that set defaults. Possible that an automated test would be a really good thing. On a single machine, you can easily create a table, set some data, and then select that data on the slave.

**Comment by sameersbn** [comment]
ok.. I didn't know that the PR was updated. GitHub does not send notifications on such event. I will test it out and merge the PR as soon as possible.

On Wednesday 08 July 2015 09:11 AM, Charles Pritchard wrote:
> @sameersbn Yes - my merge went awrong: I missed the chunk of code that set defaults. Possible that an automated test would be a really good thing. On a single machine, you can easily create a table, set some data, and then select that data on the slave.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/sameersbn/docker-postgresql/pull/27#issuecomment-119412197
>

-- 
--
Sameer Naik

w: http://www.damagehead.com e: sameer@damagehead.com

**Comment by Downchuck** [comment]
Thanks again for the extra work, apologies about the time it has taken for extra rounds of testing.

**Comment by sameersbn** [comment]
@Downchuck I am merging this right now. However I will remove the `PSQL_ARCHIVEMODE` stuff before pushing to gitlab because I an not able to test it out. It error out for me. Maybe I am not using it right, but will hold off this change until i have some clarity.

I hope it is ok.

**Comment by sameersbn** [comment]
@Downchuck Thanks for the contribution.
